src/blog/understanding-ecmascript-part-2.md

```diff
@@ -0,0 +1,305 @@
+
+For example:
+
+```javascript
+const o1 = {'foo' : 2511};
+const o2 = {};
+console.log(o2.foo); // undefined
+o2.__proto__ = o1;
+console.log(o2.foo); // 2511
+```
+
+## Where's the prototype walk defined?
+
+Let's try to find out where this behavior is defined. A good place to start is a list of [Object Internal Methods](https://tc39.es/ecma262/#sec-object-internal-methods-and-internal-slots).
+
+There's both `[[GetOwnProperty]]` and `[[Get]]` &mdash; now we're interested in the version that isn't restricted to _own_ properties, so we'll go with `[[Get]]`.
+
+Unfortunately, the [Property Descriptor specification type](https://tc39.es/ecma262/#sec-property-descriptor-specification-type) also has a field called `[[Get]]`, so while browsing the spec for `[[Get]]`, we need to carefully distinguish between the two independent usages.
+
+The internal method `[[Get]]` delegates to `OrdinaryGet`:
+
+> [`[[Get]] ( P, Receiver )`](https://tc39.es/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-get-p-receiver)
+>
+> When the `[[Get]]` internal method of `O` is called with property key `P` and ECMAScript language value `Receiver`, the following steps are taken:
+>
+> 1. Return `? OrdinaryGet(O, P, Receiver)`.
+
+`OrdinaryGet` is defined like this:
+
+> [`OrdinaryGet ( O, P, Receiver )`](https://tc39.es/ecma262/#sec-ordinaryget)
+>
+> When the abstract operation `OrdinaryGet` is called with Object `O`, property key `P`, and ECMAScript language value `Receiver`, the following steps are taken:
+>
+> 1. Assert: `IsPropertyKey(P)` is `true`.
+> 2. Let `desc` be `? O.[[GetOwnProperty]](P)`.
+> 3. If `desc` is `undefined`, then
+> a. Let `parent` be `? O.[[GetPrototypeOf]]()`.
+> b. If `parent` is `null`, return `undefined`.
+> c. Return `? parent.[[Get]](P, Receiver)`.
+> 4. If `IsDataDescriptor(desc)` is `true`, return `desc.[[Value]]`.
+> 5. Assert: `IsAccessorDescriptor(desc)` is `true`.
+> 6. Let `getter` be `desc.[[Get]]`.
+> 7. If `getter` is `undefined`, return `undefined`.
+> 8. Return `? Call(getter, Receiver)`.
+
+The prototype chain walk is inside step 3: if we don't find the property as an own property, we recurse into the prototype's `[[Get]]` method.
+
+## Where's the Receiver coming from?
+
+`OrdinaryGet` also passes the original `Receiver` throughout the recursion, unchanged. The `Receiver` is only used if the property is an accessor property, in which case it's passed as the **this value** when calling the getter function.
+
+Let's find out where the `Receiver` is coming from!
+
+Searching for places where `[[Get]]` is called we find an abstract operation `GetValue` which operates on References. Reference is a specification type, consisting of a base value (in this case, an Object), the reference name (in this case, a String), and a strict mode flag (bool).
```

Comment:
maybe `(in this case, a String)` => `(in this case, the string 'foo')`?

Reply:
I got more and more confused below. Could you make the example super explicit here? 
In our example, the reference consists of the following values: base = o1 (or o2? I don't know); reference name = `foo`, strict mode flag = true.